File: library/src/main/java/com/owncloud/android/lib/resources/albums/ReadAlbumItemsRemoteOperation.java
/*
 * Nextcloud Android Library
 *
 * SPDX-FileCopyrightText: 2025 TSI-mc <surinder.kumar@t-systems.com>
 * SPDX-License-Identifier: AGPL-3.0-or-later
 */

package com.owncloud.android.lib.resources.albums;

import com.nextcloud.common.SessionTimeOut;
import com.nextcloud.common.SessionTimeOutKt;
import com.owncloud.android.lib.common.OwnCloudClient;
import com.owncloud.android.lib.common.network.WebdavUtils;
import com.owncloud.android.lib.common.operations.RemoteOperation;
import com.owncloud.android.lib.common.operations.RemoteOperationResult;
import com.owncloud.android.lib.common.utils.Log_OC;
import com.owncloud.android.lib.common.utils.WebDavFileUtils;
import com.owncloud.android.lib.resources.files.model.RemoteFile;

import org.apache.commons.httpclient.HttpStatus;
import org.apache.jackrabbit.webdav.DavConstants;
import org.apache.jackrabbit.webdav.MultiStatus;
import org.apache.jackrabbit.webdav.client.methods.PropFindMethod;

import java.util.ArrayList;
import java.util.List;

public class ReadAlbumItemsRemoteOperation extends RemoteOperation<List<RemoteFile>> {

    private static final String TAG = ReadAlbumItemsRemoteOperation.class.getSimpleName();
    private final String mRemotePath;
    private final SessionTimeOut sessionTimeOut;

    public ReadAlbumItemsRemoteOperation(String mRemotePath) {
        this(mRemotePath, SessionTimeOutKt.getDefaultSessionTimeOut());
    }

    public ReadAlbumItemsRemoteOperation(String mRemotePath, SessionTimeOut sessionTimeOut) {
        this.mRemotePath = mRemotePath;
        this.sessionTimeOut = sessionTimeOut;
    }

    protected RemoteOperationResult<List<RemoteFile>> run(OwnCloudClient client) {
        RemoteOperationResult<List<RemoteFile>> result = null;
        PropFindMethod query = null;
        String url = client.getBaseUri() + "/remote.php/dav/photos/" + client.getUserId() + "/albums" + WebdavUtils.encodePath(mRemotePath);
        try {
            // remote request
            query = new PropFindMethod(url,
                WebdavUtils.getAllPropSet(), // PropFind Properties
                DavConstants.DEPTH_1);
            int status = client.executeMethod(query, sessionTimeOut.getReadTimeOut(), sessionTimeOut.getConnectionTimeOut());

            // check and process response
            boolean isSuccess = (status == HttpStatus.SC_MULTI_STATUS || status == HttpStatus.SC_OK);

            if (isSuccess) {
                // get data from remote folder
                MultiStatus dataInServer = query.getResponseBodyAsMultiStatus();
                ArrayList<RemoteFile> mFolderAndFiles = new WebDavFileUtils().readAlbumData(dataInServer, client);

                // Result of the operation
                result = new RemoteOperationResult<>(true, query);
                // Add data to the result
                if (result.isSuccess()) {
                    result.setResultData(mFolderAndFiles);
                }
            } else {
                // synchronization failed
                client.exhaustResponse(query.getResponseBodyAsStream());
                result = new RemoteOperationResult<>(false, query);
            }
        } catch (Exception e) {
            result = new RemoteOperationResult<>(e);
        } finally {
            if (query != null) {
                query.releaseConnection(); // let the connection available for other methods
            }

            if (result == null) {
                result = new RemoteOperationResult(new Exception("unknown error"));
                Log_OC.e(TAG, "Synchronized " + mRemotePath + ": failed");
            } else {
                if (result.isSuccess()) {
                    Log_OC.i(TAG, "Synchronized " + mRemotePath + ": " + result.getLogMessage());
                } else {
                    if (result.isException()) {
                        Log_OC.e(TAG, "Synchronized " + mRemotePath + ": " + result.getLogMessage(),
                            result.getException());
                    } else {
                        Log_OC.e(TAG, "Synchronized " + mRemotePath + ": " + result.getLogMessage());
                    }
                }
            }
        }

        return result;
    }
}

File: library/src/main/java/com/owncloud/android/lib/resources/albums/ReadAlbumItemsRemoteOperation.kt
/*
 * Nextcloud Android Library
 *
 * SPDX-FileCopyrightText: 2026 Alper Ozturk <alper.ozturk@nextcloud.com>
 * SPDX-FileCopyrightText: 2025 TSI-mc <surinder.kumar@t-systems.com>
 * SPDX-License-Identifier: MIT
 */
package com.owncloud.android.lib.resources.albums

import com.nextcloud.common.SessionTimeOut
import com.nextcloud.common.defaultSessionTimeOut
import com.owncloud.android.lib.common.OwnCloudClient
import com.owncloud.android.lib.common.network.WebdavUtils
import com.owncloud.android.lib.common.operations.RemoteOperation
import com.owncloud.android.lib.common.operations.RemoteOperationResult
import com.owncloud.android.lib.common.utils.Log_OC
import com.owncloud.android.lib.common.utils.WebDavFileUtils
import com.owncloud.android.lib.resources.files.model.RemoteFile
import org.apache.jackrabbit.webdav.DavConstants
import org.apache.jackrabbit.webdav.client.methods.PropFindMethod

class ReadAlbumItemsRemoteOperation
    @JvmOverloads
    constructor(
        private val remotePath: String,
        private val sessionTimeOut: SessionTimeOut = defaultSessionTimeOut
    ) : RemoteOperation<List<RemoteFile>>() {
        @Deprecated("Deprecated in Java")
        @Suppress("TooGenericExceptionCaught")
        override fun run(client: OwnCloudClient): RemoteOperationResult<List<RemoteFile>> {
            var propFind: PropFindMethod? = null
            val result =
                try {
                    propFind =
                        PropFindMethod(
                            client.albumUri(remotePath),
                            WebdavUtils.getAlbumItemPropSet(),
                            DavConstants.DEPTH_1
                        )
                    val status =
                        client.executeMethod(propFind, sessionTimeOut.readTimeOut, sessionTimeOut.connectionTimeOut)

                    if (isMultiStatusOrOk(status)) {
                        RemoteOperationResult<List<RemoteFile>>(true, propFind).apply {
                            resultData = WebDavFileUtils().readAlbumData(propFind.responseBodyAsMultiStatus, client)
                        }
                    } else {
                        client.exhaustResponse(propFind.responseBodyAsStream)
                        RemoteOperationResult(false, propFind)
                    }
                } catch (e: OutOfMemoryError) {
                    Log_OC.e(TAG, "Not enough memory to read the content of $remotePath", e)
                    RemoteOperationResult(RemoteOperationResult.ResultCode.OUT_OF_MEMORY)
                } catch (e: Exception) {
                    RemoteOperationResult(e)
                } finally {
                    propFind?.releaseConnection()
                }

            return result.also { log(it) }
        }

        @Suppress("DEPRECATION")
        private fun log(result: RemoteOperationResult<List<RemoteFile>>) {
            val message = "Synchronized $remotePath : ${result.logMessage}"
            when {
                result.isSuccess -> Log_OC.i(TAG, message)
                result.isException -> Log_OC.e(TAG, message, result.exception)
                else -> Log_OC.e(TAG, message)
            }
        }

        companion object {
            private val TAG: String = ReadAlbumItemsRemoteOperation::class.java.simpleName
        }
    }
